src/realm/sync/subscriptions.hpp

```diff
@@ -0,0 +1,287 @@
+/*************************************************************************
+ *
+ * Copyright 2021 Realm, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#pragma once
+
+#include "realm/db.hpp"
+#include "realm/list.hpp"
+#include "realm/obj.hpp"
+#include "realm/query.hpp"
+#include "realm/timestamp.hpp"
+#include "realm/util/future.hpp"
+#include "realm/util/optional.hpp"
+
+#include <string_view>
+
+namespace realm::sync {
+
+class SubscriptionSet;
+class SubscriptionStore;
+
+// A Subscription represents a single query that may be OR'd with other queries on the same object class to be
+// send to the server in a QUERY or IDENT message.
+class Subscription {
+public:
+    // Returns the timestamp of when this subscription was originally created.
+    Timestamp created_at() const;
+
+    // Returns the timestamp of the last time this subscription was updated by calling update_query.
+    Timestamp updated_at() const;
+
+    // Returns the name of the subscription that was set when it was created. If no name was set, this
+    // returns a default-constructed std::string_view.
+    StringData name() const;
+
+    // Returns the name of the object class of the query for this subscription.
+    StringData object_class_name() const;
+
+    // Returns a stringified version of the query associated with this subscription.
+    StringData query_string() const;
+
+    // Replaces the query/object class of this query with another one.
+    //
+    // If another Subscription exists in the SubscriptionSet containing this subscription with the same
+    // query, then this will throw a std::runtime_error.
+    void update_query(const Query& query_obj);
```

Comment:
Do we need to have this on the `Subscription` class? I would imagine SDKs would only update queries by calling `SubscriptionSet::insert`.

Reply:
I mean, I don't know? I would imagine the opposite. I figured it'd be useful to be able to see what subscriptions are in the subscription set without modifying them and to separate conceptually the container from the value it contains?

Reply:
Since we already have the `insert` on `SubscriptionSet`, I'd rather not expose multiple ways of doing the same thing. If someone wants to update an existing subscription, they can pass its name to `SubscriptionSet::insert` with the `update` flag.